lib/backend/pgbk/pgbk.go

```diff
@@ -0,0 +1,580 @@
+		pool:   pool,
+		buf:    backend.NewCircularBuffer(),
+		cancel: cancel,
+	}
+
+	if err := b.setupAndMigrate(ctx); err != nil {
+		b.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if !cfg.DisableExpiry {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.backgroundExpiry(ctx)
+		}()
+	}
+
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+		b.backgroundChangeFeed(ctx)
+	}()
+
+	return b, nil
+}
+
+// Backend is a PostgreSQL-backed [backend.Backend].
+type Backend struct {
+	cfg  Config
+	log  logrus.FieldLogger
+	pool *pgxpool.Pool
+	buf  *backend.CircularBuffer
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func (b *Backend) Close() error {
+	b.cancel()
+	b.wg.Wait()
+	b.buf.Close()
+	b.pool.Close()
+	return nil
+}
+
+var schemas = []string{
+	`CREATE TABLE kv (
+		key bytea NOT NULL,
+		value bytea NOT NULL,
+		expires timestamptz,
+		revision uuid NOT NULL,
+		CONSTRAINT kv_pkey PRIMARY KEY (key)
+	);
+	CREATE INDEX kv_expires_idx ON kv (expires) WHERE expires IS NOT NULL;`,
+}
+
+// setupAndMigrate sets up the database schema, applying the migrations in the
+// [schemas] slice starting from the first non-applied one.
+func (b *Backend) setupAndMigrate(ctx context.Context) error {
+	var version int32
+	var migrateErr error
+
+	// this is split off from the rest because we might not have permissions to
+	// CREATE TABLE, which is checked even if the table exists
+	if _, err := pgcommon.RetryIdempotent(ctx, b.log, func() (struct{}, error) {
+		_, err := b.pool.Exec(ctx,
+			`CREATE TABLE IF NOT EXISTS backend_version (
+				version integer PRIMARY KEY CHECK (version > 0),
+				created timestamptz NOT NULL DEFAULT now()
+			)`, pgx.QueryExecModeExec,
+		)
+		return struct{}{}, trace.Wrap(err)
+	}); err != nil {
+		// the very first SELECT in the next transaction will fail, we don't
+		// need anything higher than debug here
+		b.log.WithError(err).Debug("Failed to confirm the existence of the backend_version table.")
+	}
+
+	const idempotent = true
+	if err := pgcommon.RetryTx(ctx, b.log, b.pool, pgx.TxOptions{
+		IsoLevel:   pgx.Serializable,
+		AccessMode: pgx.ReadWrite,
+	}, idempotent, func(tx pgx.Tx) error {
+		if err := tx.QueryRow(ctx,
+			"SELECT COALESCE(max(version), 0) FROM backend_version",
+			pgx.QueryExecModeExec,
+		).Scan(&version); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if int(version) > len(schemas) {
+			migrateErr = trace.BadParameter("unsupported schema version %v", version)
+			return nil
+		}
+
+		if int(version) == len(schemas) {
+			return nil
+		}
+
+		for _, s := range schemas[version:] {
+			if _, err := tx.Exec(ctx, s, pgx.QueryExecModeExec); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		if _, err := tx.Exec(ctx,
+			"INSERT INTO backend_version (version) VALUES ($1)",
+			pgx.QueryExecModeExec, len(schemas),
+		); err != nil {
+			return trace.Wrap(err)
+		}
+
+		return nil
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if migrateErr != nil {
+		return migrateErr
```

Comment:
```suggestion
		return trace.Wrap(migrateErr)
```

Reply:
but it was already a trace error 😭

fixed in 04043cc1cffa246aa6e4404488bfe02d92030a65